src/proto/batch/mod.rs

```diff
@@ -0,0 +1,338 @@
+/// let child_cb = Job::builder("clean_up").build();
+///
+/// let parent_batch = Batch::builder("Image recognition and analysis workload".to_string()).with_complete_callback(parent_cb);
+/// let child_batch = Batch::builder("Image recognition workload".to_string()).with_success_callback(child_cb);
+///
+/// let mut parent = prod.start_batch(parent_batch)?;
+/// parent.add(parent_job1)?;
+/// parent.add(parent_job2)?;
+/// let mut child = parent.start_batch(child_batch)?;
+/// child.add(child_job1)?;
+/// child.add(child_job2)?;
+///
+/// child.commit()?;
+/// parent.commit()?;
+///
+/// # Ok::<(), Error>(())
+/// ```
+///
+/// In the example above, there is a single level nesting, but you can nest those batches as deep as you wish,
+/// effectively building a pipeline this way, since the Faktory guarantees that callback jobs will not be queued unless
+/// the batch gets committed.
+///
+/// You can retieve the batch status using a [`tracker`](struct.Tracker.html):
+/// ```no_run
+/// # use faktory::Error;
+/// # use faktory::{Producer, Job, Batch, Tracker};
+/// let mut prod = Producer::connect(None)?;
+/// let job = Job::builder("image").build();
+/// let cb_job = Job::builder("clean_up").build();
+/// let b = Batch::builder("Description...".to_string()).with_complete_callback(cb_job);
+///
+/// let mut b = prod.start_batch(b)?;
+/// let bid = b.id().to_string();
+/// b.add(job)?;
+/// b.commit()?;
+///
+/// let mut t = Tracker::connect(None).unwrap();
+/// let s = t.get_batch_status(bid).unwrap().unwrap();
+/// assert_eq!(s.total, 1);
+/// assert_eq!(s.pending, 1);
+/// assert_eq!(s.description, Some("Description...".into()));
+/// assert_eq!(s.complete_callback_state, ""); // has not been queued;
+/// # Ok::<(), Error>(())
+/// ```
+#[derive(Serialize, Debug, Builder)]
+#[builder(
+    custom_constructor,
+    setter(into),
+    build_fn(name = "try_build", private)
+)]
+pub struct Batch {
+    #[serde(skip_serializing_if = "Option::is_none")]
+    #[builder(setter(skip))]
+    parent_bid: Option<String>,
+
+    /// Batch description for Faktory WEB UI.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub description: Option<String>,
+
+    /// On success callback.
+    ///
+    /// This job will be queued by the Faktory server provided
+    /// all the jobs belonging to this batch have been executed successfully.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    #[builder(setter(custom))]
+    pub(crate) success: Option<Job>,
+
+    /// On complete callback.
+    ///
+    /// This job will be queue by the Faktory server after all the jobs
+    /// belonging to this batch have been executed, even if one/some/all
+    /// of the workers have failed.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    #[builder(setter(custom))]
+    pub(crate) complete: Option<Job>,
+}
+
+impl Batch {
+    /// Create a new `BatchBuilder`.
+    pub fn builder(description: impl Into<Option<String>>) -> BatchBuilder {
```

Comment:
I think I would recommend we simply disallow name-less batches, and make this argument `impl Into<String>`. It'll make the interface nicer to use ultimately, and get rid of a bunch of `.to_string()` calls.

Reply:
Alternatively, don't take any arguments to `builder`, and instead just have people use the builder to set the name if they so wish.

Reply:
Whichever way you go, I'd do the same thing for `BatchBuilder::new` below.

Reply:
I like the latter option better, just to avoid making it opinionated

Reply:
Iterating on it again, I believe the `description` setter should be custom and _really_ should have an `impl Into<String>` so that on the surface it's:
```rust
let batch = Batch::builder()
    .description("Batch description")
    .with_complete_callback(job_cb);
```